### ISSUE TITLE
Does dropout work in the Recognition/Torch code?

### ISSUE THREAD

**Issue by marlowinnovations** [deepinsight/insightface#2434]
I'm using a Resnet backbone.  Currently the dropout ratio is set to 0.0 .. I tried various values all the way up to 0.99 and it didn't seem to have any affect at all on the training.  Is the code not functioning properly or does dropout not have any affect the way the network is designed?

**Comment by Sunshine066** [comment]
Is you detect or recognition models?

**Comment by Sunshine066** [comment]
> - class IResNet(nn.Module):
> -     fc_scale = 7 * 7
> -     def __init__(self,
> -                  block, layers, dropout=0, num_features=512, zero_init_residual=False,
> -                  groups=1, width_per_group=64, replace_stride_with_dilation=None, fp16=False):
> -         super(IResNet, self).__init__()
> -         self.extra_gflops = 0.0
> -         self.fp16 = fp16
> -         self.inplanes = 64
> -         self.dilation = 1
> -         if replace_stride_with_dilation is None:
> -             replace_stride_with_dilation = [False, False, False]
> -         if len(replace_stride_with_dilation) != 3:
> -             raise ValueError("replace_stride_with_dilation should be None "
> -                              "or a 3-element tuple, got {}".format(replace_stride_with_dilation))
> -         self.groups = groups
> -         self.base_width = width_per_group
> -         self.conv1 = nn.Conv2d(3, self.inplanes, kernel_size=3, stride=1, padding=1, bias=False)
> -         self.bn1 = nn.BatchNorm2d(self.inplanes, eps=1e-05)
> -         self.prelu = nn.PReLU(self.inplanes)
> -         self.layer1 = self._make_layer(block, 64, layers[0], stride=2)
> -         self.layer2 = self._make_layer(block,
> -                                        128,
> -                                        layers[1],
> -                                        stride=2,
> -                                        dilate=replace_stride_with_dilation[0])
> -         self.layer3 = self._make_layer(block,
> -                                        256,
> -                                        layers[2],
> -                                        stride=2,
> -                                        dilate=replace_stride_with_dilation[1])
> -         self.layer4 = self._make_layer(block,
> -                                        512,
> -                                        layers[3],
> -                                        stride=2,
> -                                        dilate=replace_stride_with_dilation[2])
> -         self.bn2 = nn.BatchNorm2d(512 * block.expansion, eps=1e-05,)
> -         self.dropout = nn.Dropout(p=dropout, inplace=True)
> -         self.fc = nn.Linear(512 * block.expansion * self.fc_scale, num_features)
> -         self.features = nn.BatchNorm1d(num_features, eps=1e-05)
> -         nn.init.constant_(self.features.weight, 1.0)
> -         self.features.weight.requires_grad = False
> - 
> -         for m in self.modules():
> -             if isinstance(m, nn.Conv2d):
> -                 nn.init.normal_(m.weight, 0, 0.1)
> -             elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
> -                 nn.init.constant_(m.weight, 1)
> -                 nn.init.constant_(m.bias, 0)
> - 
> -         if zero_init_residual:
> -             for m in self.modules():
> -                 if isinstance(m, IBasicBlock):
> -                     nn.init.constant_(m.bn2.weight, 0)


Are you sure you have successfully modified this parameter of dropout=0

**Comment by Sunshine066** [comment]
If you modify it successfully, it must have some effect, I have tested

**Comment by marlowinnovations** [comment]
Ok thanks.. I'm pretty sure I tried changing that dropout parameter in the init function, but I'll try it again and let you know.  Just confirm it should be a float value right?  So 0.50 for 50% dropout probability?

**Comment by Sunshine066** [comment]
> Ok thanks.. I'm pretty sure I tried changing that dropout parameter in the init function, but I'll try it again and let you know. Just confirm it should be a float value right? So 0.50 for 50% dropout probability?

You are right.

**Comment by marlowinnovations** [comment]
I set dropout to 0.90 in the init function just to get an exaggerated setting.  You can see in this loss chart, pink is dropout 0.90 and blue is dropout of 0.0.  It's nearly identical.

![image](https://github.com/deepinsight/insightface/assets/46036056/291c30df-ef8c-4ffb-a256-3cd5080fad00)

![image](https://github.com/deepinsight/insightface/assets/46036056/9b3b75ad-2e79-45a2-a724-108f457bcd3b)

**Comment by Sunshine066** [comment]
> I set dropout to 0.90 in the init function just to get an exaggerated setting. You can see in this loss chart, pink is dropout 0.90 and blue is dropout of 0.0. It's nearly identical.
> 
> ![image](https://user-images.githubusercontent.com/46036056/269987950-291c30df-ef8c-4ffb-a256-3cd5080fad00.png)
> 
> ![image](https://user-images.githubusercontent.com/46036056/269987527-9b3b75ad-2e79-45a2-a724-108f457bcd3b.png)

My guess:


You simply changed the default value of dropout here, but here it is

if name == "r18":

return iresnet18(False, **kwargs)

kwargs contains the incoming value of dropout, so if you change the default value here, it's equivalent to not changing it.


You can try printing it.

And how did you create this curve?

self.dropout = nn.Dropout(p=dropout, inplace=True)


The p here is already defined in IResNet initialization, and during training,

def forward(self, x):

with torch.cuda.amp.autocast(self.fp16):

x = self.conv1(x)

x = self.bn1(x)

x = self.prelu(x)

x = self.layer1(x)

x = self.layer2(x)

x = self.layer3(x)

x = self.layer4(x)

x = self.bn2(x)

x = torch.flatten(x, 1)

x = self.dropout(x)

x = self.fc(x.float() if self.fp16 else x)

x = self.features(x)

return x

Even if you change p, since IResNet has determined it, x here = self.dropout(x)  will not change

**Comment by Sunshine066** [comment]
I suggest you tell me the code logic of how you got this diagram, and how you modified the code logic of the dropout coefficient so that I can better solve your problem

**Comment by marlowinnovations** [comment]
Thanks @Sunshine066 .  So irestnet.py is left untouched

![image](https://github.com/deepinsight/insightface/assets/46036056/7553e590-9c05-4a5b-841c-fb846a6c571e)

In train_v2.py, I changed dropout=0.0 to dropout=0.90 and trained for 10000 iterations.. then set dropout to 0.0 and trained for 10000 iterations.. that is where I got that loss curve I posted previously.  The loss curve was no different between the two settings which is why I felt like it wasn't working.  I would think a drop out of 0.90 would have a significant impact on the training.

![image](https://github.com/deepinsight/insightface/assets/46036056/7159ddf4-6b66-47fc-b663-82e8b3726700)
